### PR TITLE
fix #153 target only the uri schemes we understand

### LINKED ISSUE #153: LS DocumentSelector should confine itself to schemes it can handle
Currently, LS will try to provide its services to any document that is set to use one of these language modes:
https://github.com/intersystems/language-server/blob/87fad7b5d59d36158ef5b687166a81f1f2acea50/client/src/extension.ts#L50-L55

But if the document uri's scheme isn't `file`, `objectscript`, `isfs` or `isfs-readonly` this can result in failures. For instance, if a Serenji user also has the vscode-objectscript and languageserver extensions active, VS Code defaults the language mode for a .cls document to `objectscript-class` rather than `serenji.cls` despite the document having been opened using the `serenji` scheme rather than the `isfs` sscheme.

As things stand, LS handles the document but if the user hovers over a symbol that causes LS to need to contact the server for information it mishandles the serenji:// uri.

I will submit a PR.

### PR DESCRIPTION
**Pull request by gjsjohnmurray**
This PR resolves #153

Submitted as a draft because I can't build LS to test it. @isc-bsaviano please assist.

### CODE REVIEW

**Comment by isc-bsaviano** [comment]
@gjsjohnmurray Why did you change the import? It's supposed to be "vscode-languageserver/node" since 7.0.0. What issues are you running into with the build?

**Comment by gjsjohnmurray** [comment]
> 
> 
> @gjsjohnmurray Why did you change the import? It's supposed to be "vscode-languageserver/node" since 7.0.0. What issues are you running into with the build?

Until I changed that, my VS Code reported this problem at line 20 of client/src/extension.ts

image

This remained even after I re-ran `npm install` from the root folder:
```
C:\Users\XXX\Documents\GitHub\YYY\language-server>npm install

> language-server@1.1.4 postinstall
> cd client && npm install && cd ../server && npm install && cd ..


up to date, audited 13 packages in 866ms

found 0 vulnerabilities

up to date, audited 53 packages in 806ms

found 0 vulnerabilities

up to date, audited 262 packages in 4s

found 0 vulnerabilities

C:\Users\XXX\Documents\GitHub\YYY\language-server>
```

**Comment by gjsjohnmurray** [comment]
I have reverted that change. The problem shows again, but since I'm not equipped to build it this doesn't affect me.

**Comment by isc-bsaviano** [comment]
@gjsjohnmurray I'm not seeing that problem on my end and I know that Isabelle hasn't had any issues with testing her refactoring stuff. If you delete all 3 of your node_modules folders and do an npm install again, does that help?

**Comment by gjsjohnmurray** [comment]
@isc-bsaviano that cleared the problem, but the `npm install` reported a problem during the `node-gyp rebuild`

```
npm ERR! code 1
npm ERR! path C:\Users\JohnM\Documents\GitHub\gjsjohnmurray\language-server\server\node_modules\ref-struct-napi\node_modules\ref-napi
npm ERR! command failed
npm ERR! command C:\WINDOWS\system32\cmd.exe /d /s /c node-gyp rebuild
npm ERR! Building the projects in this solution one at a time. To enable parallel build, please add the "/m" switch.
npm ERR!   nothing.c
npm ERR! C:\Users\JohnM\Documents\GitHub\gjsjohnmurray\language-server\server\node_modules\node-addon-api\src\nothing.c : fatal error C1083: Cannot open compiler generated file: 'C:\Users\JohnM\Documents\GitHub\gjsjohnmurray\language-server\server\node_modules\node-addon-api\src\Release\obj\nothing\nothing.obj': Permission denied [C:\Users\JohnM\Documents\GitHub\gjsjohnmurray\language-server\server\node_modules\node-addon-api\src\nothing.vcxproj]
npm ERR! gyp info it worked if it ends with ok
npm ERR! gyp info using node-gyp@7.1.2
npm ERR! gyp info using node@12.18.3 | win32 | x64
npm ERR! gyp info find Python using Python version 2.7.15 found at "C:\Users\JohnM\.windows-build-tools\python27\python.exe"
npm ERR! gyp info find VS using VS2017 (15.9.28307.1300) found at:
npm ERR! gyp info find VS "C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools"
npm ERR! gyp info find VS run with --verbose for detailed information
npm ERR! gyp info spawn C:\Users\JohnM\.windows-build-tools\python27\python.exe
npm ERR! gyp info spawn args [
npm ERR! gyp info spawn args   'C:\\Users\\JohnM\\AppData\\Roaming\\npm\\node_modules\\npm\\node_modules\\node-gyp\\gyp\\gyp_main.py',
npm ERR! gyp info spawn args   'binding.gyp',
npm ERR! gyp info spawn args   '-f',
npm ERR! gyp info spawn args   'msvs',
npm ERR! gyp info spawn args   '-I',
npm ERR! gyp info spawn args   'C:\\Users\\JohnM\\Documents\\GitHub\\gjsjohnmurray\\language-server\\server\\node_modules\\ref-struct-napi\\node_modules\\ref-napi\\build\\config.gypi',
npm ERR! gyp info spawn args   '-I',
npm ERR! gyp info spawn args   'C:\\Users\\JohnM\\AppData\\Roaming\\npm\\node_modules\\npm\\node_modules\\node-gyp\\addon.gypi',
npm ERR! gyp info spawn args   '-I',
npm ERR! gyp info spawn args   'C:\\Users\\JohnM\\AppData\\Local\\node-gyp\\Cache\\12.18.3\\include\\node\\common.gypi',
npm ERR! gyp info spawn args   '-Dlibrary=shared_library',
npm ERR! gyp info spawn args   '-Dvisibility=default',
npm ERR! gyp info spawn args   '-Dnode_root_dir=C:\\Users\\JohnM\\AppData\\Local\\node-gyp\\Cache\\12.18.3',
npm ERR! gyp info spawn args   '-Dnode_gyp_dir=C:\\Users\\JohnM\\AppData\\Roaming\\npm\\node_modules\\npm\\node_modules\\node-gyp',
npm ERR! gyp info spawn args   '-Dnode_lib_file=C:\\\\Users\\\\JohnM\\\\AppData\\\\Local\\\\node-gyp\\\\Cache\\\\12.18.3\\\\<(target_arch)\\\\node.lib',
npm ERR! gyp info spawn args   '-Dmodule_root_dir=C:\\Users\\JohnM\\Documents\\GitHub\\gjsjohnmurray\\language-server\\server\\node_modules\\ref-struct-napi\\node_modules\\ref-napi',
npm ERR! gyp info spawn args   '-Dnode_engine=v8',
npm ERR! gyp info spawn args   '--depth=.',
npm ERR! gyp info spawn args   '--no-parallel',
npm ERR! gyp info spawn args   '--generator-output',
npm ERR! gyp info spawn args   'C:\\Users\\JohnM\\Documents\\GitHub\\gjsjohnmurray\\language-server\\server\\node_modules\\ref-struct-napi\\node_modules\\ref-napi\\build',
npm ERR! gyp info spawn args   '-Goutput_dir=.'
npm ERR! gyp info spawn args ]
npm ERR! gyp info spawn C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\MSBuild\15.0\Bin\MSBuild.exe
npm ERR! gyp info spawn args [
npm ERR! gyp info spawn args   'build/binding.sln',
npm ERR! gyp info spawn args   '/clp:Verbosity=minimal',
npm ERR! gyp info spawn args   '/nologo',
npm ERR! gyp info spawn args   '/p:Configuration=Release;Platform=x64'
npm ERR! gyp info spawn args ]
npm ERR! gyp ERR! build error
npm ERR! gyp ERR! stack Error: `C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\MSBuild\15.0\Bin\MSBuild.exe` failed with exit code: 1
npm ERR! gyp ERR! stack     at ChildProcess.onExit (C:\Users\JohnM\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\lib\build.js:194:23)
npm ERR! gyp ERR! stack     at ChildProcess.emit (events.js:315:20)
npm ERR! gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:275:12)
npm ERR! gyp ERR! System Windows_NT 10.0.19042
npm ERR! gyp ERR! command "C:\\Program Files\\nodejs\\node.exe" "C:\\Users\\JohnM\\AppData\\Roaming\\npm\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
npm ERR! gyp ERR! cwd C:\Users\JohnM\Documents\GitHub\gjsjohnmurray\language-server\server\node_modules\ref-struct-napi\node_modules\ref-napi
npm ERR! gyp ERR! node -v v12.18.3
npm ERR! gyp ERR! node-gyp -v v7.1.2
npm ERR! gyp ERR! not ok

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\JohnM\AppData\Local\npm-cache\_logs\2021-04-28T13_26_42_410Z-debug.log
npm ERR! code 1
npm ERR! path C:\Users\JohnM\Documents\GitHub\gjsjohnmurray\language-server
npm ERR! command failed
npm ERR! command C:\WINDOWS\system32\cmd.exe /d /s /c cd client && npm install && cd ../server && npm install && cd ..

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\JohnM\AppData\Local\npm-cache\_logs\2021-04-28T13_26_42_440Z-debug.log
PS C:\Users\JohnM\Documents\GitHub\gjsjohnmurray\language-server>
```

**Comment by isc-bsaviano** [comment]
@gjsjohnmurray That's coming from the node-ffi stuff that Andy uses to call out to the parsers. I haven't seen that build error on Mac

**Comment by isc-bsaviano** [comment]
@gjsjohnmurray I just cloned the LS repo on my windows machine and using node 14.15.0, I was able to successfully run `npm install` and build all of the dependencies.

**Comment by gjsjohnmurray** [comment]
I just tried updating my Node JS from 12.18.3 to 14.16.1 and installing Visual Studio 2019 Community Edition so it could be used instead of Visual Studio 2017 Build Tools, but basically the same outcome:

```
PS C:\Users\JohnM\Documents\GitHub\gjsjohnmurray\language-server> npm install

> language-server@1.1.4 postinstall
> cd client && npm install && cd ../server && npm install && cd ..


up to date, audited 13 packages in 782ms

found 0 vulnerabilities
npm ERR! code 1
npm ERR! path C:\Users\JohnM\Documents\GitHub\gjsjohnmurray\language-server\server\node_modules\ref-struct-napi\node_modules\ref-napi
npm ERR! command failed
npm ERR! command C:\WINDOWS\system32\cmd.exe /d /s /c node-gyp rebuild
npm ERR! Building the projects in this solution one at a time. To enable parallel build, please add the "-m" switch.
npm ERR!   nothing.c
npm ERR! C:\Users\JohnM\Documents\GitHub\gjsjohnmurray\language-server\server\node_modules\node-addon-api\src\nothing.c : fatal error C1083: Cannot open compiler generated file: 'C:\Users\JohnM\Documents\GitHub\gjsjohnmurray\language-server\server\node_modules\node-addon-api\src\Release\obj\nothing\nothing.obj': Permission denied [C:\Users\JohnM\Documents\GitHub\gjsjohnmurray\language-server\server\node_modules\node-addon-api\src\nothing.vcxproj]
npm ERR! gyp info it worked if it ends with ok
npm ERR! gyp info using node-gyp@7.1.2
npm ERR! gyp info using node@14.16.1 | win32 | x64
npm ERR! gyp info find Python using Python version 2.7.15 found at "C:\Users\JohnM\.windows-build-tools\python27\python.exe"
npm ERR! gyp info find VS using VS2019 (16.9.31205.134) found at:
npm ERR! gyp info find VS "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community"
npm ERR! gyp info find VS run with --verbose for detailed information
npm ERR! gyp info spawn C:\Users\JohnM\.windows-build-tools\python27\python.exe
npm ERR! gyp info spawn args [
npm ERR! gyp info spawn args   'C:\\Users\\JohnM\\AppData\\Roaming\\npm\\node_modules\\npm\\node_modules\\node-gyp\\gyp\\gyp_main.py',
npm ERR! gyp info spawn args   'binding.gyp',
npm ERR! gyp info spawn args   '-f',
npm ERR! gyp info spawn args   'msvs',
npm ERR! gyp info spawn args   '-I',
npm ERR! gyp info spawn args   'C:\\Users\\JohnM\\Documents\\GitHub\\gjsjohnmurray\\language-server\\server\\node_modules\\ref-struct-napi\\node_modules\\ref-napi\\build\\config.gypi',
npm ERR! gyp info spawn args   '-I',
npm ERR! gyp info spawn args   'C:\\Users\\JohnM\\AppData\\Roaming\\npm\\node_modules\\npm\\node_modules\\node-gyp\\addon.gypi',
npm ERR! gyp info spawn args   '-I',
npm ERR! gyp info spawn args   'C:\\Users\\JohnM\\AppData\\Local\\node-gyp\\Cache\\14.16.1\\include\\node\\common.gypi',
npm ERR! gyp info spawn args   '-Dlibrary=shared_library',
npm ERR! gyp info spawn args   '-Dvisibility=default',
npm ERR! gyp info spawn args   '-Dnode_root_dir=C:\\Users\\JohnM\\AppData\\Local\\node-gyp\\Cache\\14.16.1',
npm ERR! gyp info spawn args   '-Dnode_gyp_dir=C:\\Users\\JohnM\\AppData\\Roaming\\npm\\node_modules\\npm\\node_modules\\node-gyp',
npm ERR! gyp info spawn args   '-Dnode_lib_file=C:\\\\Users\\\\JohnM\\\\AppData\\\\Local\\\\node-gyp\\\\Cache\\\\14.16.1\\\\<(target_arch)\\\\node.lib',
npm ERR! gyp info spawn args   '-Dmodule_root_dir=C:\\Users\\JohnM\\Documents\\GitHub\\gjsjohnmurray\\language-server\\server\\node_modules\\ref-struct-napi\\node_modules\\ref-napi',
npm ERR! gyp info spawn args   '-Dnode_engine=v8',
npm ERR! gyp info spawn args   '--depth=.',
npm ERR! gyp info spawn args   '--no-parallel',
npm ERR! gyp info spawn args   '--generator-output',
npm ERR! gyp info spawn args   'C:\\Users\\JohnM\\Documents\\GitHub\\gjsjohnmurray\\language-server\\server\\node_modules\\ref-struct-napi\\node_modules\\ref-napi\\build',
npm ERR! gyp info spawn args   '-Goutput_dir=.'
npm ERR! gyp info spawn args ]
npm ERR! gyp info spawn C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Current\Bin\MSBuild.exe
npm ERR! gyp info spawn args [
npm ERR! gyp info spawn args   'build/binding.sln',
npm ERR! gyp info spawn args   '/clp:Verbosity=minimal',
npm ERR! gyp info spawn args   '/nologo',
npm ERR! gyp info spawn args   '/p:Configuration=Release;Platform=x64'
npm ERR! gyp info spawn args ]
npm ERR! gyp ERR! build error
npm ERR! gyp ERR! stack Error: `C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Current\Bin\MSBuild.exe` failed with exit code: 1
npm ERR! gyp ERR! stack     at ChildProcess.onExit (C:\Users\JohnM\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\lib\build.js:194:23)
npm ERR! gyp ERR! stack     at ChildProcess.emit (events.js:315:20)
npm ERR! gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:277:12)
npm ERR! gyp ERR! System Windows_NT 10.0.19042
npm ERR! gyp ERR! command "C:\\Program Files\\nodejs\\node.exe" "C:\\Users\\JohnM\\AppData\\Roaming\\npm\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
npm ERR! gyp ERR! cwd C:\Users\JohnM\Documents\GitHub\gjsjohnmurray\language-server\server\node_modules\ref-struct-napi\node_modules\ref-napi
npm ERR! gyp ERR! node -v v14.16.1
npm ERR! gyp ERR! node-gyp -v v7.1.2
npm ERR! gyp ERR! not ok

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\JohnM\AppData\Local\npm-cache\_logs\2021-04-28T15_13_01_304Z-debug.log
npm ERR! code 1
npm ERR! path C:\Users\JohnM\Documents\GitHub\gjsjohnmurray\language-server
npm ERR! command failed
npm ERR! command C:\WINDOWS\system32\cmd.exe /d /s /c cd client && npm install && cd ../server && npm install && cd ..

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\JohnM\AppData\Local\npm-cache\_logs\2021-04-28T15_13_01_329Z-debug.log
PS C:\Users\JohnM\Documents\GitHub\gjsjohnmurray\language-server>
```

File: client/src/extension.ts [hunk 1]
```diff
@@ -47,16 +48,35 @@ export async function activate(context: ExtensionContext) {
...
+		'objectscript-class',
+		'objectscript-csp',
+		'objectscript-macros',
 	];
 
+	// The uri schemes we handle those languages for
+	const targetSchemes = [
+		'isfs',
+		'isfs-readonly',
+		'objectscripts',
```
Comment (isc-bsaviano) [inline client/src/extension.ts:63]: This is a typo. The scheme is 'objectscript'. Also, do we need to include the 'objectscriptxml' schema from the main extension?

Comment (gjsjohnmurray) [reply client/src/extension.ts]: Well spotted re the typo. And I've added `objectscriptxml` as it won't hurt to have it there.

**Comment by gjsjohnmurray** [comment]
> 
> 
> @gjsjohnmurray It looks like the error was a permissions issue:
> `fatal error C1083: Cannot open compiler generated file: 'C:\Users\JohnM\Documents\GitHub\gjsjohnmurray\language-server\server\node_modules\node-addon-api\src\Release\obj\nothing\nothing.obj': Permission denied`

Interesting. When I restored server\node_modules from the trash the problem went away. So perhaps some kind of bootstrapping issue, not being able to 'npm install' for server in a clean folder.

**Comment by isc-bsaviano** [comment]
@gjsjohnmurray Does the LS start up in debug mode now? The correct launch configuration is `Client + Server`

**Comment by gjsjohnmurray** [comment]
Launching with that config creates two rows in CALL STACK. The Launch Client one stays around, but the Attach to Server one disappears when this alert appears:

image

But `vsce package` was able to build the VSIX so I am now testing with that.

**Comment by gjsjohnmurray** [comment]
@isc-bsaviano testing with  the VSIX I verified it now only handles the specified schemes.

**Comment by isc-bsaviano** [comment]
@gjsjohnmurray That error pops up sometimes but if you just cancel it and go to the extension host, everything should work fine. I'm not sure where that's coming from.

**Pull request merged**